Safari is not using the intended AudioSession

grorg opened the pull request:
#### e0ce48fe5c2e1b20a22f817e081face321c7c541
<pre>
Safari is not using the intended AudioSession
<a href="https://bugs.webkit.org/show_bug.cgi?id=255681">https://bugs.webkit.org/show_bug.cgi?id=255681</a>
rdar://107095456

Reviewed by Jer Noble.

WebKit never sets the AudioSession mode to anything other than &quot;default&quot;, which
is not appropriate for video playback, especially in a spatial audio environment.
Fix this by exposing a way for the Media system to select an AudioSession mode,
and set it to MoviePlayback when needed.

* LayoutTests/media/audio-session-category-at-most-recent-playback-expected.txt:
* LayoutTests/media/audio-session-category-at-most-recent-playback.html:
* LayoutTests/media/audio-session-category-expected.txt:
* LayoutTests/media/audio-session-category.html:

* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h: Expose the new mode constant.
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm:

* Source/WebCore/html/HTMLMediaElement.cpp: Save the mode so that tests can query it.
(WebCore::HTMLMediaElement::playPlayer):
* Source/WebCore/html/HTMLMediaElement.h:
(WebCore::HTMLMediaElement::modeAtMostRecentPlayback const):

* Source/WebCore/platform/audio/AudioSession.h: Add a new type AudioSession::Mode that can be
set alongside the Category value.
(WTF::LogArgument&lt;WebCore::AudioSession::Mode&gt;::toString):
* Source/WebCore/platform/audio/AudioSession.cpp:
(WebCore::AudioSession::setCategory):
(WebCore::AudioSession::setCategoryOverride):
(WebCore::AudioSession::mode const):
(WebCore::convertEnumerationToString):

* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::updateSessionState): Set the mode to VideoPlayback here
if `hasAudibleAudioOrVideoMediaType` is true.

* Source/WebCore/platform/audio/ios/AudioSessionIOS.h:
* Source/WebCore/platform/audio/ios/AudioSessionIOS.mm:
(WebCore::AudioSessionIOS::setCategory): Actually send the mode down to the AVAudioSession.
(WebCore::AudioSessionIOS::mode const):

* Source/WebCore/platform/audio/mac/AudioSessionMac.h:
* Source/WebCore/platform/audio/mac/AudioSessionMac.mm:
(WebCore::AudioSessionMac::setCategory): Accept the parameter, but we don&apos;t set anything on Mac.

* Source/WebCore/platform/mock/MockRealtimeAudioSource.cpp:
(WebCore::MockRealtimeAudioSource::startProducingData): Update call to setCategory.

* Source/WebCore/testing/Internals.cpp: Expose a way to test the mode.
(WebCore::Internals::audioSessionMode const):
(WebCore::Internals::modeAtMostRecentPlayback const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
(WebKit::RemoteAudioSessionProxy::setCategory):
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.h:
(WebKit::RemoteAudioSessionProxy::mode const):
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.cpp:
(WebKit::RemoteAudioSession::setCategory):
(WebKit::RemoteAudioSession::mode const):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.h:
* Source/WebKit/WebProcess/Speech/SpeechRecognitionRealtimeMediaSourceManager.cpp:
    Update the call to setCategory to accept Mode.

* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxyManager.cpp:
(WebKit::RemoteAudioSessionProxyManager::updateCategory):
    Iterate through the proxies to test if mode has been set.

Canonical link: <a href="https://commits.webkit.org/263201@main">https://commits.webkit.org/263201@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/199171a1db6795145f6a26150f93c1972265beb7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3918 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4010 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4121 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5356 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4178 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3895 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4094 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/4000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/3991 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3966 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/4162 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/3534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5205 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/1665 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/3509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/5547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/3485 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/3569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/4977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3972 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/4000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3491 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/3509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3539 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/446 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3770 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
```diff
@@ -180,13 +180,16 @@
+    } else if (hasAudibleAudioOrVideoMediaType) {
         category = AudioSession::CategoryType::MediaPlayback;
-    else if (webAudioCount)
+        mode = AudioSession::Mode::MoviePlayback;
```
We should only set `mode = AudioSession::Mode::MoviePlayback` when there's a non-zero count of _audible_ `VideoAudio` sessions.

jernoble commented on Source/WebCore/platform/audio/AudioSession.cpp:
```diff
@@ -168,7 +168,7 @@ void AudioSession::setCategoryOverride(CategoryType category)
     if (category != CategoryType::None)
-        setCategory(category, RouteSharingPolicy::Default);
+        setCategory(category, mode(), RouteSharingPolicy::Default);
 }
```
This seems dangerous. Maybe we should instead just pass in Default in this case?

grorg committed:
Safari is not using the intended AudioSession
https://bugs.webkit.org/show_bug.cgi?id=255681
rdar://107095456

Reviewed by Jer Noble.

WebKit never sets the AudioSession mode to anything other than "default", which
is not appropriate for video playback, especially in a spatial audio environment.
Fix this by exposing a way for the Media system to select an AudioSession mode,
and set it to MoviePlayback when needed.

* LayoutTests/media/audio-session-category-at-most-recent-playback-expected.txt:
* LayoutTests/media/audio-session-category-at-most-recent-playback.html:
* LayoutTests/media/audio-session-category-expected.txt:
* LayoutTests/media/audio-session-category.html:

* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h: Expose the new mode constant.
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm:

* Source/WebCore/html/HTMLMediaElement.cpp: Save the mode so that tests can query it.
(WebCore::HTMLMediaElement::playPlayer):
* Source/WebCore/html/HTMLMediaElement.h:
(WebCore::HTMLMediaElement::modeAtMostRecentPlayback const):

* Source/WebCore/platform/audio/AudioSession.h: Add a new type AudioSession::Mode that can be
set alongside the Category value.
(WTF::LogArgument<WebCore::AudioSession::Mode>::toString):
* Source/WebCore/platform/audio/AudioSession.cpp:
(WebCore::AudioSession::setCategory):
(WebCore::AudioSession::setCategoryOverride):
(WebCore::AudioSession::mode const):
(WebCore::convertEnumerationToString):

* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::updateSessionState): Set the mode to VideoPlayback here
if `hasAudibleAudioOrVideoMediaType` is true.

* Source/WebCore/platform/audio/ios/AudioSessionIOS.h:
* Source/WebCore/platform/audio/ios/AudioSessionIOS.mm:
(WebCore::AudioSessionIOS::setCategory): Actually send the mode down to the AVAudioSession.
(WebCore::AudioSessionIOS::mode const):

* Source/WebCore/platform/audio/mac/AudioSessionMac.h:
* Source/WebCore/platform/audio/mac/AudioSessionMac.mm:
(WebCore::AudioSessionMac::setCategory): Accept the parameter, but we don't set anything on Mac.

* Source/WebCore/platform/mock/MockRealtimeAudioSource.cpp:
(WebCore::MockRealtimeAudioSource::startProducingData): Update call to setCategory.

* Source/WebCore/testing/Internals.cpp: Expose a way to test the mode.
(WebCore::Internals::audioSessionMode const):
(WebCore::Internals::modeAtMostRecentPlayback const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
(WebKit::RemoteAudioSessionProxy::setCategory):
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.h:
(WebKit::RemoteAudioSessionProxy::mode const):
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.cpp:
(WebKit::RemoteAudioSession::setCategory):
(WebKit::RemoteAudioSession::mode const):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.h:
* Source/WebKit/WebProcess/Speech/SpeechRecognitionRealtimeMediaSourceManager.cpp:
    Update the call to setCategory to accept Mode.

* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxyManager.cpp:
(WebKit::RemoteAudioSessionProxyManager::updateCategory):
    Iterate through the proxies to test if mode has been set.

Canonical link: https://commits.webkit.org/263201@main
